src/engraving/libmscore/tempotext.cpp

```diff
@@ -308,51 +446,41 @@ void TempoText::undoChangeProperty(Pid id, const QVariant& v, PropertyFlags ps)
 
 void TempoText::updateTempo()
 {
-    // cache regexp, they are costly to create
-    static QHash<QString, QRegularExpression> regexps;
-    static QHash<QString, QRegularExpression> regexps2;
-    QString s = plainText();
-    s.replace(",", ".");
-    s.replace("<sym>space</sym>", " ");
-    for (const TempoPattern& pa : tp) {
-        QRegularExpression re;
-        if (!regexps.contains(pa.pattern)) {
-            re = QRegularExpression(QString("%1\\s*=\\s*(\\d+[.]{0,1}\\d*)\\s*").arg(pa.pattern));
-            regexps[pa.pattern] = re;
-        }
-        re = regexps.value(pa.pattern);
-        QRegularExpressionMatch match = re.match(s);
-        if (match.hasMatch()) {
-            QStringList sl = match.capturedTexts();
-            if (sl.size() == 2) {
-                qreal nt = qreal(sl[1].toDouble()) * pa.f;
-                if (nt != _tempo) {
-                    undoChangeProperty(Pid::TEMPO, QVariant(qreal(sl[1].toDouble()) * pa.f), propertyFlags(Pid::TEMPO));
-                    _relative = 1.0;
-                    _isRelative = false;
-                    updateScore();
-                }
-                break;
-            }
-        } else {
-            for (const TempoPattern& pa2 : tp) {
-                QString key = QString("%1_%2").arg(pa.pattern, pa2.pattern);
-                QRegularExpression re2;
-                if (!regexps2.contains(key)) {
-                    re2 = QRegularExpression(QString("%1\\s*=\\s*%2\\s*").arg(pa.pattern, pa2.pattern));
-                    regexps2[key] = re2;
-                }
-                re2 = regexps2.value(key);
-                QRegularExpressionMatch match2 = re2.match(s);
-                if (match2.hasMatch()) {
-                    _relative = pa2.f / pa.f;
-                    _isRelative = true;
-                    updateRelative();
-                    updateScore();
-                    return;
-                }
+    static const float quarterNotePlayback = 1.f / 60.f;
+
+    static const QRegularExpression bpmExpression("^[\\[,\\(]?[ ]*(?<note>[dwhqest].*) *= *(?<bpm>\\d+)[\\],\\)]?$");
```

Comment:
"dwhqest" occurs 8 times in this file and I would think is moderately likely to need modifying in future, better to use a named constant.